torch/_functorch/vmap.py

```diff
@@ -258,26 +259,95 @@ def _get_name(func: Callable):
     return repr(func)
 
 
-def vmap_impl(func, in_dims, out_dims, randomness, chunk_size, *args, **kwargs):
+def vmap_impl(
+    func, in_dims, out_dims, randomness, chunk_size, chunk_with_scan, *args, **kwargs
+):
     lazy_load_decompositions()
     _check_out_dims_is_int_or_int_pytree(out_dims, func)
     batch_size, flat_in_dims, flat_args, args_spec = _process_batched_inputs(
         in_dims, args, func
     )
 
     if chunk_size is not None:
-        chunks_flat_args = _get_chunked_inputs(
-            flat_args, flat_in_dims, batch_size, chunk_size
-        )
-        return _chunked_vmap(
-            func,
-            flat_in_dims,
-            chunks_flat_args,
-            args_spec,
-            out_dims,
-            randomness,
-            **kwargs,
-        )
+        if chunk_with_scan:
+            if batch_size % chunk_size != 0:
+                # TODO: support padding
+                raise NotImplementedError(
+                    f"vmap_impl: batch_size ({batch_size}) must be divisible by chunk_size ({chunk_size})"
+                )
+
+            num_chunks = batch_size // chunk_size
+
+            def _normalize_dims(in_dims):
+                if isinstance(in_dims, int) or in_dims is None:
+                    assert len(args) == 1
+                    in_dims = (in_dims,)
+                return in_dims
+
+            in_dims: tuple[Optional[int], ...] = _normalize_dims(in_dims)
```

Comment:
there's a broadcast function somewhere for broadcast semantics. can do in follow-up (shouldn't be difficult)